internal/buffer/buffer.go

```diff
@@ -889,6 +898,33 @@ func (b *Buffer) UpdateRules() {
 					break
 				}
 			}
+
+			if b.SyntaxDef == nil {
+				// search for the default file in the runtime files
+				for _, f := range config.ListRuntimeFiles(config.RTSyntax) {
+					if f.Name() == "default" {
+						data, err := f.Data()
+						if err != nil {
+							screen.TermMessage("Error loading syntax file " + f.Name() + ": " + err.Error())
+							continue
+						}
+
+						file, err := highlight.ParseFile(data)
+						if err != nil {
+							screen.TermMessage("Error parsing syntax file " + f.Name() + ": " + err.Error())
+							continue
+						}
+
+						syndef, err := highlight.ParseDef(file, header)
+						if err != nil {
+							screen.TermMessage("Error parsing syntax file " + f.Name() + ": " + err.Error())
+							continue
+						}
+						b.SyntaxDef = syndef
+						break
+					}
+				}
+			}
```

Comment:
How about de-duplicating the code? The size of `UpdateRules()` is getting scary.

Reply:
I started the refactoring in #3206. This PR will benefit of it too in the moment of necessary rebase.

Reply:
My work on #3208 made me realize that such refactoring is probably is not an easy thing to do, unfortunately. The logic in `UpdateRules()` is quite subtle, it's not obvious to me how to abstract any important pieces of it into separate functions without loss of functionality and correctness.

Perhaps we should approach this from a different direction: instead of trying to restructure the code, just add some helpers to the syntax parser API, e.g. `ParseDefFromFile()` (which would internally just call `ParseFile()` and then `ParseDef()`), maybe also `FindRuntimeFile()` and `FindRealRuntimeFile()` or something like that, to make the code of `UpdateRules()` at least shorter and a bit easier to read.

Reply:
> [...] just add some helpers to the syntax parser API, e.g. ParseDefFromFile() (which would internally just call ParseFile() and then ParseDef()), maybe also FindRuntimeFile() and FindRealRuntimeFile() or something like that, to make the code of UpdateRules() at least shorter and a bit easier to read.

Yep, this will reduce the duplication of load/parses and logs in that way that they aren't spread all over `UpdateRules()`.
I will take care of this in the moment #3208 is merged, because then the whole change needs a rebase and I will drop some lines I did.